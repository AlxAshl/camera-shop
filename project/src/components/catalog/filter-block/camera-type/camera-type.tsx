import { useEffect } from 'react';
import { useSearchParams } from 'react-router-dom';
import { CategoryFilter, TypeFilter, URLParams } from '../../../../const';


export function CameraType() {

  const [searchParams, setSearchParams] = useSearchParams();
  const videoCam: HTMLInputElement | null = document.querySelector('input[name="videocamera"]');
  const photoCam: HTMLInputElement | null = document.querySelector('input[name="photocamera"]');
  const typeFilters = searchParams.getAll(URLParams.Type);
  const categoryFilters = searchParams.getAll(URLParams.Category);

  const handleTypeInputChange = (type: string) => {
    if (!typeFilters.includes(type)) {
      searchParams.append(URLParams.Type, type);
    }
    else {
      const typeFilterArray = typeFilters.filter((entry) => entry !== type);
      searchParams.delete(URLParams.Type);
      typeFilterArray.forEach((entry) => {
        searchParams.append(URLParams.Type, entry);
      });
    }
    setSearchParams(searchParams);
  };

  useEffect(()=>{
    if(categoryFilters.includes(CategoryFilter.Video) && (typeFilters.includes(TypeFilter.Film) || typeFilters.includes(TypeFilter.Instant)) && !categoryFilters.includes(CategoryFilter.Camera)) {
      const typeFilterArray = typeFilters.filter((entry) => entry !== TypeFilter.Film).filter((entry) => entry !== TypeFilter.Instant);
      searchParams.delete(URLParams.Type);
      typeFilterArray.forEach((entry) => {
        searchParams.append(URLParams.Type, entry);
      });
      setSearchParams(searchParams);
    }
  },[categoryFilters]);

  return(
    <fieldset className="catalog-filter__block">
      <legend className="title title--h5">Тип камеры</legend>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='digital-test' name="digital" checked={typeFilters.includes(TypeFilter.Digital)} onChange={() => handleTypeInputChange(TypeFilter.Digital)}/>
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Цифровая</span>
        </label>
      </div>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='film-test' name="film" checked={typeFilters.includes(TypeFilter.Film)} disabled={videoCam?.checked && !photoCam?.checked} onChange={() =>handleTypeInputChange(TypeFilter.Film)}/>
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Плёночная</span>
        </label>
      </div>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='snapshot-test' name="snapshot" checked={typeFilters.includes(TypeFilter.Instant)} disabled={videoCam?.checked && !photoCam?.checked} onChange={() =>handleTypeInputChange(TypeFilter.Instant)}/>
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Моментальная</span>
        </label>
      </div>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='collection-test' name="collection" checked={typeFilters.includes(TypeFilter.Collectible)} onChange={() =>handleTypeInputChange(TypeFilter.Collectible)}/>
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Коллекционная</span>
        </label>
      </div>
    </fieldset>
  );
}
