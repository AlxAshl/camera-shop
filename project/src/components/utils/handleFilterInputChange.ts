import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';

export const handleFilterInputChange = (filterName: string, filter: string[], action: ActionCreatorWithNonInferrablePayload) => {
  const mutableFilter = Object.assign([], filter);
  if(filter.includes(filterName)) {
    const updatedFilter = mutableFilter.filter((entry) => entry !== filterName);
    store.dispatch(action(updatedFilter));
  }
  else {
    mutableFilter.push(filterName);
    store.dispatch(action(mutableFilter));
  }
};
