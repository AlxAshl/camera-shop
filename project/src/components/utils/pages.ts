import { PAGE_LIMIT } from '../../const';

function getPagesCount(cameraCount: number): number {
  return Math.ceil(cameraCount / PAGE_LIMIT);
}

export default getPagesCount;
