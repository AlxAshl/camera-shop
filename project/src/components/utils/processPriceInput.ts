import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';
import { ProductType } from '../../types/product';

export const setMinInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMinPriceInput;
    const newMin = productsArray.find((product) => product.price <= Number(input));
    newMin?.price === undefined
      ? newMinPriceInput = productsArray[productsArray.length - 1].price
      : newMinPriceInput = newMin?.price;
    store.dispatch(action([String(newMinPriceInput)]));
  }
};

export const setMaxInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMaxPriceInput;
    const newMax = productsArray.find((product) => product.price >= Number(input));
    newMax?.price === undefined
      ? newMaxPriceInput = productsArray[productsArray.length - 1].price
      : newMaxPriceInput = newMax?.price;
    store.dispatch(action([String(newMaxPriceInput)]));
  }
};
