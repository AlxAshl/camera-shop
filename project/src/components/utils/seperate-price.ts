export function seperatePrice(priceValue: number) {
  let returnedPrice = String(priceValue);
  if(returnedPrice.length > 6) {
    returnedPrice = `${returnedPrice.slice(0, 1)} ${returnedPrice.slice(1, 4)} ${returnedPrice.slice(-3)}`;
  }
  if (returnedPrice.length > 3) {
    returnedPrice = `${returnedPrice.slice(-0, -3)} ${returnedPrice.slice(-3)}`;
  }
  return returnedPrice;
}
