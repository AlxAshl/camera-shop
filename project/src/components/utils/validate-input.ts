export function validateInput(target: (EventTarget & HTMLInputElement) | (EventTarget & HTMLTextAreaElement)) {

  if (target.parentElement?.parentElement && target.name === 'rating' && (Number(target.value) > 0)) {
    if (!target.parentElement.parentElement.parentElement?.className.includes('is-valid')) {
      target.parentElement.parentElement.parentElement?.classList.add('is-valid');
      target.parentElement.parentElement.parentElement?.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name !== 'review' && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length > 0) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string'){
    if(target.value.length === 0) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length >= 5) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name === 'review' && target.value.length < 5) {
    if (!target.parentElement.parentElement.className.includes('is-invalid')) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

}
