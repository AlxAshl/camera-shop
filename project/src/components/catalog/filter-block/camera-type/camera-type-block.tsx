import { useEffect } from 'react';
import { CategoryFilter, TypeFilter } from '../../../../const';
import { useAppDispatch } from '../../../../hooks/useAppDispatch';
import { useAppSelector } from '../../../../hooks/useAppSelector';
import { typeFilterSetter } from '../../../../store/filters-process/filters-process';
import { getFilters } from '../../../../store/filters-process/selectors';
import { handleFilterInputChange } from '../../../utils/handleFilterInputChange';


export function CameraTypeBlock() {

  const videoCam: HTMLInputElement | null = document.querySelector('input[name="videocamera"]');
  const photoCam: HTMLInputElement | null = document.querySelector('input[name="photocamera"]');
  const {Type, Category} = useAppSelector(getFilters);
  const dispatch = useAppDispatch();

  useEffect(()=>{
    if(Category.includes(CategoryFilter.Video) && (Type.includes(TypeFilter.Film) || Type.includes(TypeFilter.Instant)) && !Category.includes(CategoryFilter.Camera)) {
      const typeFilterArray = Type.filter((entry) => entry !== TypeFilter.Film).filter((entry) => entry !== TypeFilter.Instant);
      dispatch(typeFilterSetter(typeFilterArray));
    }
  },[Category]);

  return(
    <fieldset className="catalog-filter__block">
      <legend className="title title--h5">Тип камеры</legend>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='digital-test' name="digital" checked={Type.includes(TypeFilter.Digital)} onChange={
            () => handleFilterInputChange(TypeFilter.Digital, Type, typeFilterSetter)
          }
          />
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Цифровая</span>
        </label>
      </div>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='film-test' name="film" checked={Type.includes(TypeFilter.Film)} disabled={videoCam?.checked && !photoCam?.checked} onChange={
            () => handleFilterInputChange(TypeFilter.Film, Type, typeFilterSetter)
          }
          />
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Плёночная</span>
        </label>
      </div>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='snapshot-test' name="snapshot" checked={Type.includes(TypeFilter.Instant)} disabled={videoCam?.checked && !photoCam?.checked} onChange={
            () => handleFilterInputChange(TypeFilter.Instant, Type, typeFilterSetter)
          }
          />
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Моментальная</span>
        </label>
      </div>
      <div className="custom-checkbox catalog-filter__item">
        <label>
          <input type="checkbox" data-testid='collection-test' name="collection" checked={Type.includes(TypeFilter.Collectible)} onChange={
            () =>handleFilterInputChange(TypeFilter.Collectible, Type, typeFilterSetter)
          }
          />
          <span className="custom-checkbox__icon"></span>
          <span className="custom-checkbox__label">Коллекционная</span>
        </label>
      </div>
    </fieldset>
  );
}
