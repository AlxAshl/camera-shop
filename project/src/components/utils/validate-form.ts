import { reviewDataType } from '../../types/review';
import { validateInput } from './validate-input';


export function validateForm(data: reviewDataType) {
  let validationStatus = true;
  Object.entries(data).forEach(([key, value]) =>{
    if (key !== 'review') {
      if(value === '') {
        const failedInput = document.querySelector(`input[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
        validationStatus = false;
      }
    }
    if(key === 'review') {
      if(String(value).length < 5) {
        const failedInput = document.querySelector(`textarea[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
      }
    }
    if(key === 'rating') {
      if(value < 1) {
        const failedInput = document.querySelector(`input[name=${key}]`);
        if(failedInput && failedInput.parentElement?.parentElement){
          failedInput.parentElement.parentElement.parentElement?.classList.add('is-invalid');
        }
        validationStatus = false;
      }
    }
  });
  return validationStatus;
}

