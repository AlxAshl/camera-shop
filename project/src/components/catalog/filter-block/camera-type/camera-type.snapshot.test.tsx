import { renderWithProviders } from '../../../../test/utils/render-with-redux';
import { CameraType } from './camera-type';

test('Render CameraType', () => {
  const { asFragment } = renderWithProviders(<CameraType />);
  expect(asFragment()).toMatchSnapshot();
});
