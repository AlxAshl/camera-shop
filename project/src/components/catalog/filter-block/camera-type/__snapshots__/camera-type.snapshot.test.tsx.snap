// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CameraType 1`] = `
<DocumentFragment>
  <fieldset
    class="catalog-filter__block"
  >
    <legend
      class="title title--h5"
    >
      Тип камеры
    </legend>
    <div
      class="custom-checkbox catalog-filter__item"
    >
      <label>
        <input
          data-testid="digital-test"
          name="digital"
          type="checkbox"
        />
        <span
          class="custom-checkbox__icon"
        />
        <span
          class="custom-checkbox__label"
        >
          Цифровая
        </span>
      </label>
    </div>
    <div
      class="custom-checkbox catalog-filter__item"
    >
      <label>
        <input
          data-testid="film-test"
          name="film"
          type="checkbox"
        />
        <span
          class="custom-checkbox__icon"
        />
        <span
          class="custom-checkbox__label"
        >
          Плёночная
        </span>
      </label>
    </div>
    <div
      class="custom-checkbox catalog-filter__item"
    >
      <label>
        <input
          data-testid="snapshot-test"
          name="snapshot"
          type="checkbox"
        />
        <span
          class="custom-checkbox__icon"
        />
        <span
          class="custom-checkbox__label"
        >
          Моментальная
        </span>
      </label>
    </div>
    <div
      class="custom-checkbox catalog-filter__item"
    >
      <label>
        <input
          data-testid="collection-test"
          name="collection"
          type="checkbox"
        />
        <span
          class="custom-checkbox__icon"
        />
        <span
          class="custom-checkbox__label"
        >
          Коллекционная
        </span>
      </label>
    </div>
  </fieldset>
</DocumentFragment>
`;
